flang/lib/Lower/Allocatable.cpp

```diff
@@ -348,44 +767,180 @@ void Fortran::lower::genDeallocateStmt(
   errorManagement.lower(converter, loc, statExpr, errMsgExpr);
   for (const auto &allocateObject :
        std::get<std::list<Fortran::parser::AllocateObject>>(stmt.t)) {
-    const auto &symbol = unwrapSymbol(allocateObject);
-    auto boxAddr = converter.getSymbolAddress(symbol);
-    if (!boxAddr)
-      TODO("Allocatable type not lowered yet");
-    // TODO use return stat for error recovery
-    genAllocatableDeallocate(builder, loc, boxAddr, errorManagement.hasStat,
-                             errorManagement.errMsgBoxAddr,
-                             errorManagement.sourceFile,
-                             errorManagement.sourceLine);
+    auto box = genMutableBoxValue(converter, loc, allocateObject);
+    genDeallocate(builder, loc, box, errorManagement);
   }
 }
 
-void Fortran::lower::genAllocatableInit(
-    Fortran::lower::AbstractConverter &converter,
-    const Fortran::lower::pft::Variable &var, mlir::Value boxAddress) {
-  auto loc = converter.genLocation(var.getSymbol().name());
-  auto &builder = converter.getFirOpBuilder();
-  auto declType = var.getSymbol().GetType();
-  if (!declType)
-    TODO("Is this possible ?");
-  if (auto intrinsic = declType->AsIntrinsic()) {
-    if (intrinsic->category() == Fortran::common::TypeCategory::Character) {
-      TODO("character alloctable init");
-    } else {
-      auto i32ty = builder.getIntegerType(32);
-      int catCode = static_cast<int>(intrinsic->category());
-      auto cat = builder.createIntegerConstant(loc, i32ty, catCode);
-      auto kindExpr = Fortran::evaluate::AsGenericExpr(
-          Fortran::common::Clone(intrinsic->kind()));
-      Fortran::lower::StatementContext stmtCtx;
-      auto kind = fir::getBase(converter.genExprValue(kindExpr, stmtCtx));
-      auto rank =
-          builder.createIntegerConstant(loc, i32ty, var.getSymbol().Rank());
-      auto corank = builder.createIntegerConstant(loc, i32ty, 0);
-      genAllocatableInitIntrinsic(builder, loc, boxAddress, cat, kind, rank,
-                                  corank);
+//===----------------------------------------------------------------------===//
+// MutableBoxValue creation implementation
+//===----------------------------------------------------------------------===//
+
+mlir::Value
+Fortran::lower::createUnallocatedBox(Fortran::lower::FirOpBuilder &builder,
+                                     mlir::Location loc, mlir::Type boxType,
+                                     mlir::ValueRange nonDeferredParams) {
+  auto heapType = boxType.dyn_cast<fir::BoxType>().getEleTy();
+  auto type = fir::dyn_cast_ptrEleTy(heapType);
+  auto eleTy = type;
+  if (auto seqType = eleTy.dyn_cast<fir::SequenceType>())
+    eleTy = seqType.getEleTy();
+  if (eleTy.isa<fir::RecordType>())
+    mlir::emitError(loc, "TODO: Derived type allocatable initialization");
```

Comment:
Yes, because that is not putting compilation in an undefined states, just producing an executable with bad semantics. So I assume it is nicer to not completely  block Steve compilation testing on this.